src/Enum.php

```diff
@@ -2,425 +2,141 @@
 
 namespace Spatie\Enum;
 
-use ArgumentCountError;
 use BadMethodCallException;
-use JsonSerializable;
 use ReflectionClass;
-use ReflectionMethod;
-use Spatie\Enum\Exceptions\DuplicatedIndexException;
-use Spatie\Enum\Exceptions\DuplicatedValueException;
-use Spatie\Enum\Exceptions\InvalidIndexException;
-use Spatie\Enum\Exceptions\InvalidNameException;
-use Spatie\Enum\Exceptions\InvalidValueException;
-use TypeError;
-
-abstract class Enum implements Enumerable, JsonSerializable
-{
-    /** @var array[] */
-    protected static $cache = [];
-
-    /** @var int */
-    protected $index;
 
-    /** @var string */
+/**
+ * @property-read string value
+ * @property-read string label
+ */
+abstract class Enum
+{
+    /** @var string|int */
     protected $value;
 
-    /** @var string */
-    protected $name;
-
-    /**
-     * This construct is not part of the public API and COULD change in a minor release.
-     * You SHOULD NOT use it by your own - instead you SHOULD use the make() method.
-     *
-     * @internal
-     * @see \Spatie\Enum\Enum::make()
-     *
-     * @param string|null $name
-     * @param string|null $value
-     * @param int|null $index
-     */
-    public function __construct(?string $name = null, ?string $value = null, ?int $index = null)
-    {
-        if (is_null($name) && is_null($value) && is_null($index)) {
-            ['name' => $name, 'value' => $value, 'index' => $index] = $this->resolveByStaticCall();
-        }
+    protected string $label;
 
-        if (is_null($name) || ! static::isValidName($name)) {
-            throw new InvalidNameException($name, static::class);
-        }
+    private static array $definitionCache = [];
 
-        if (is_null($value) || ! static::isValidValue($value)) {
-            throw new InvalidValueException($value, static::class);
-        }
-
-        if (is_null($index) || ! static::isValidIndex($index)) {
-            throw new InvalidIndexException($index, static::class);
-        }
-
-        $this->name = $name;
-        $this->value = $value;
-        $this->index = $index;
-    }
-
-    public function __call(string $name, array $arguments)
+    public static function toArray(): array
     {
-        if (static::startsWith($name, 'is')) {
-            return $this->isEqual(substr($name, 2));
-        }
+        $array = [];
 
-        if (static::isValidName($name)) {
-            return static::make($name);
+        foreach (static::resolveDefinition() as $definition) {
+            $array[$definition->value] = $definition->label;
```

Comment:
Right now the `toArray()` method does the inverse.
```php
[
  'value' => 'index',
]
```
After renaming this would be `label` as key and `value` as value. But I agree that this could be not this usable if the label is used, like it suggests, for more readable strings.

So open for discussion here and I understand why key and value are flipped.

Reply:
It of course depends on when you'd want to user `toArray`, my assumption is to provide select fields in forms to select an enum value. In that case `$key => $value` is the most logical approach, since that's the way a select box would need it.

Reply:
I think that I would use `array_keys($enum->toArray())` for a select input and run the array values through `trans()` (Laravel) or anything.
But because that would be still possible everything's fine for me. I'm just thinking about methods that only return values, labels - like the `getValues()`, `getIndices()` methods before. But these would "conflict" with the current `values()` and `labels()` mapping methods.